core/src/main/java/org/elasticsearch/common/joda/DateMathParser.java

```diff
@@ -190,6 +190,11 @@ private long parseMath(String mathString, long time, boolean roundUp, DateTimeZo
     private long parseDateTime(String value, DateTimeZone timeZone, boolean roundUpIfNoTime) {
         DateTimeFormatter parser = dateTimeFormatter.parser();
         if (timeZone != null) {
+            String format = dateTimeFormatter.format();
+            // Non UTC time zones are not compatible with epoch formats
+            if (timeZone != DateTimeZone.UTC && format.equals("epoch_second") || format.equals("epoch_millis")) {
```

Comment:
Shouldn't this be 
` if (timeZone != DateTimeZone.UTC && (format.equals("epoch_second") || format.equals("epoch_millis"))) {`
[This](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/operators.html) says && takes precedence over ||.